src/projections.jl

```diff
@@ -446,9 +546,96 @@ References:
 (https://stanford.edu/~boyd/papers/cone_prog_refine.html)
 by Enzo Busseti, Walaa M. Moursi, and Stephen Boyd
 """
-function projection_gradient_on_set(::DefaultDistance, v::AbstractVector{T}, ::MOI.DualExponentialCone) where {T}
+function projection_gradient_on_set(d::DefaultDistance, v::AbstractVector{T}, ::MOI.DualExponentialCone) where {T}
+    # from Moreau decomposition: x = P_K(x) + P_-K*(x)
+    return I - projection_gradient_on_set(d, -v, MOI.ExponentialCone())
+end
+
+"""
+    projection_gradient_on_set(d::DefaultDistance, v::AbstractVector{T}, ::MOI.PowerCone) where {T}
+
+derivative of projection of vector `v` on the power cone
+i.e. `K = {(x,y,z) | x^a * y^(1-a) >= |z|, x>=0, y>=0}`.
+
+References:
+* [Differential properties of Euclidean projection onto power cone]
+(https://link.springer.com/article/10.1007/s00186-015-0514-0), Theorem 3.1
+"""
+function projection_gradient_on_set(::DefaultDistance, v::AbstractVector{T}, s::MOI.PowerCone) where {T}
+    _check_dimension(v, s)
+
+    if _in_pow_cone(v, s)
+        return Matrix{T}(I, 3, 3)
+    end
+    if _in_pow_cone(-v, MOI.dual_set(s))
+        # if in polar cone Ko = -K*
+        return zeros(T, 3, 3)
+    end
+    if abs(v[3]) <= 1e-8
+        return _pow_cone_∇proj_case_3(v, s)
+    end
+
+    x, y, z = v
+    α = s.exponent
+    # # This handles the case where v[3]/norm(v) is small.
+    #   Phi(eps()) ≈ -r
+    #   Phi_prod(x, α, z, eps()) ≈ x + |x| = 2*max(x, 0)
+    #   => this gets to case 3 in the limit
+    # if (x < 0 || y < 0) && abs(z) <= 1e-2*norm(v)*(0.5 - abs(0.5 - α))
+    #     r = eps()
+    # else
+    #     r, _ = _solve_system_pow_cone(v, s)
+    # end
+
+    r, _ = _solve_system_pow_cone(v, s)
+    za = abs(z)
+    gx = sqrt(x^2 + 4*α*r*(za - r))
+    gy = sqrt(y^2 + 4*(1-α)*r*(za - r))
+    fx = 0.5*(x + gx)
+    fy = 0.5*(y + gy)
+
+    β = 1-α
+    K = -(α*x/gx + β*y/gy)
+    L = 2*(za - r) / (za + (za - 2r) * -K)
+    J_ii(w, γ, g) = 0.5 + w/2g + γ^2*(za - 2r)*r*L/g^2
+    J_ij = α*β*(za - 2r)*r*L/(gx*gy)
+    J = [
+        J_ii(x, α, gx)      J_ij                sign(z)*α*r*L/gx;
+        J_ij                J_ii(y, β, gy)      sign(z)*β*r*L/gy;
+        sign(z)*α*r*L/gx    sign(z)*β*r*L/gy    r/za*(1+K*L)
+    ]
+    return J
+end
+
+function _pow_cone_∇proj_case_3(v::AbstractVector{T}, s::MOI.PowerCone) where {T}
```

Comment:
This case (when `v[3] = 0`) is occasionally erring in the gradient of the projection. Specifically the third term is 0 or 1, but the finite difference is off. I think FiniteDifference is incorrect in this case, but I want to test with ForwardDiff to be sure. Some examples below:

```julia
v=[-5.600603812816107, 5.152780387514885, 0.0]
dΠ = [0.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 0.0]
case=3 
FD=[0.0 -1.654979043504589e-14 0.0; 0.0 1.0000000000004314 0.0; 0.0 0.0 0.024913617014897714]

v=[1.7644918703215298, -10.78198880680753, 0.0]
dΠ = [1.0 -0.0 -0.0; -0.0 0.0 -0.0; -0.0 -0.0 0.0]
case=3
FD=[0.999999999999962 0.0 0.0; -1.7440763866016674e-14 0.0 0.0; 0.0 0.0 0.016859166520080876]

v=[-2.4150791679319137, 0.1492395818565555, 0.0]
dΠ = [0.0 -0.0 -0.0; -0.0 1.0 -0.0; -0.0 -0.0 1.0]
case=3
FD=[0.0 -3.469446951953614e-17 0.0; 0.0 1.0423003484528706 0.0; -2.1797483976750334e-27 2.3007923605150116e-13 0.9999590997747578]
```

Reply:
Not sure how to effectively use ForwardDiff here because of the conditional. The [original source](https://link.springer.com/article/10.1007/s00186-015-0514-0) I used for implementation suggests that the FD estimate is incorrect here

Reply:
the conditional should be propagated by dual numbers normally without issues

Reply:
What I meant to say is that I thought ForwardDiff will always fall into
```Julia 
    if abs(v[3]) <= 1e-8
        return [max(v[1],0), max(v[2],0), 0]
    end
```
because I was getting results like
```
  v=[-7.89142639082619, 8.460077554766231, 0.0]
  dΠ = [0.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 1.0]
  case=3
  FD=[0.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 0.0]
  α=0.3
```
which seem to be incorrect according to the paper's derivation (I could be reading the paper incorrectly [eq (11)]).

However, on closer inspection of the results, this is not always the case. E.g.,
```
v=[-2.8370973767349805, 5.0006161004981875, 0.0]
  dΠ = [0.0 -0.0 -0.0; -0.0 1.0 -0.0; -0.0 -0.0 0.0]
  case=3
  FD=[0.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 1.0]
  α=0.9
```